pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,415 @@
+		if err != nil {
+			klog.ErrorS(err, "Failed to listen on ICMP")
+			return false, 0
+		}
+	}
+	defer socket.Close()
+
+	// Create a new ICMP packet
+	body := &icmp.Echo{
+		ID:   os.Getpid() & 0xffff, // Use the current process ID as the ICMP ID
+		Seq:  int(getICMPSeq()),
+		Data: []byte("HELLO-ANTREA-AGENT"),
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		klog.ErrorS(err, "Failed to marshal ICMP message")
+		return false, 0
+	}
+	rttStart := time.Now()
+	if _, err = socket.WriteTo(msgBytes, ip); err != nil {
+		klog.ErrorS(err, "Failed to write ICMP message")
+		return false, 0
+	}
+
+	// Reply message
+	msg.Type = replyType
+	msgBytes, err = msg.Marshal(nil)
+	if err != nil {
+		klog.ErrorS(err, "Failed to marshal ICMP message")
+		return false, 0
+	}
+
+	readBuffer := make([]byte, 1024)
+	deadline, _ := ctx.Deadline()
+	if err := socket.SetReadDeadline(deadline); err != nil {
+		klog.ErrorS(err, "Failed to set read deadline")
+		return false, 0
+	}
+
+	// Receive the ICMP message
+	for {
+		n, peer, err := socket.ReadFrom(readBuffer)
+		if err != nil {
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+				klog.ErrorS(err, "Timeout reading ICMP message")
+				return false, 0
+			}
+
+			klog.ErrorS(err, "Failed to read ICMP message")
+			return false, 0
+		}
+		if peer.String() != ip.String() {
+			klog.ErrorS(err, "Received ICMP message from unexpected peer")
+			return false, 0
+		}
+		if replyType == ipv6.ICMPTypeEchoReply {
+			// Clear checksum
+			readBuffer[2] = 0
+			readBuffer[3] = 0
+		}
+		if bytes.Equal(readBuffer[:n], msgBytes) {
+			rtt := time.Since(rttStart)
+
+			klog.InfoS("Received ICMP reply")
+			return true, rtt
+		}
+	}
+}
+
+func (m *MonitorTool) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Top level goroutine to handle termination
+	go func() {
+		<-stopCh
+		cancel()
+	}()
+
+	// Start the monitor loop
+	go m.nodeLatencyMonitorInformer.Informer().Run(stopCh)
+	go m.monitorLoop(ctx)
+}
+
+func (m *MonitorTool) monitorLoop(ctx context.Context) {
+	// Low level goroutine to handle ping loop
+	var innerStopCh chan struct{}
+	var cancelInnerStopCh func()
+
+	for {
+		select {
+		case <-ctx.Done():
+			// Stop the inner ping loop
+			if cancelInnerStopCh != nil {
+				cancelInnerStopCh()
+			}
+			return
+		case <-m.latencyConfigChanged:
+			// Start or stop the pingAll goroutine based on the latencyConfig
+			if m.latencyConfig.Enable {
+				// Stop previous pingAll goroutine
+				if cancelInnerStopCh != nil {
+					cancelInnerStopCh()
+				}
+
+				// Create a new stop channel for the new pingAll goroutine
+				innerStopCh = make(chan struct{})
+				cancelInnerStopCh = func() {
+					close(innerStopCh)
+				}
+
+				// Start new pingAll goroutine
+				go m.latencyStore.Run(innerStopCh)
```

Comment:
Does this mean that the latency map data will remain in memory if user deleted the monitoring CRD?

Reply:
`nodeInformer.Informer().Run(stopCh)` seems to automatically execute addFunc when a new addition is made to the cluster, so there is no need to manually synchronize the full amount of data.

Reply:
I think I'm more worried about mem leak here. Have you thought about instead of having `NewNodeLatencyMonitor` called on Agent start up if the feature is enabled, we instantiate it only if the CRD is created?

Reply:
I will think about it and try to control the lifecycle in ping mesh.

Reply:
We will clean all maps here.

```go
cancelInnerStopCh = func() {
	close(innerStopCh)

	// Clean up the latency store
	m.latencyStore.CleanUp()
}
```

Reply:
Let's take a step back here: why do we need to put node informer logic in latencyStore at all? If node events are handled in `MonitorTool` and we only interact with the latencyStore to update node IP mappings / LatencyEntries, then we won't need to couple the latencyStore with starting/stopping goroutinues??

Reply:
That make sense, I have removed them to `MonitorTool `.